src/main/cpp/jsonlayout.cpp

```diff
@@ -0,0 +1,273 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <log4cxx/logstring.h>
+#include <log4cxx/jsonlayout.h>
+#include <log4cxx/spi/loggingevent.h>
+#include <log4cxx/level.h>
+#include <log4cxx/helpers/optionconverter.h>
+#include <log4cxx/helpers/iso8601dateformat.h>
+#include <log4cxx/helpers/stringhelper.h>
+
+#include <apr_time.h>
+#include <apr_strings.h>
+#include <string.h>
+
+using namespace log4cxx;
+using namespace log4cxx::helpers;
+using namespace log4cxx::spi;
+
+IMPLEMENT_LOG4CXX_OBJECT(JSONLayout)
+
+
+JSONLayout::JSONLayout() : locationInfo(false), dateFormat()
+{
+}
+
+
+void JSONLayout::setOption(const LogString& option, const LogString& value)
+{
+	if (StringHelper::equalsIgnoreCase(option,
+			LOG4CXX_STR("LOCATIONINFO"), LOG4CXX_STR("locationinfo")))
+	{
+		setLocationInfo(OptionConverter::toBoolean(value, false));
+	}
+}
+void JSONLayout::format(LogString& output,
+	const spi::LoggingEventPtr& event,
+	Pool& p) const
+{
+
+	output.append("{ ");
+
+	appendQuotedEscapedString(output, "timestamp");
+	output.append(": ");
+	LogString timestamp;
+	dateFormat.format(timestamp, event->getTimeStamp(), p);
+	appendQuotedEscapedString(output, timestamp);
+	output.append(", ");
+
+	appendQuotedEscapedString(output, "level");
+	output.append(": ");
+	LogString level;
+	event->getLevel()->toString(level);
+	appendQuotedEscapedString(output, level);
+	output.append(", ");
+
+	appendQuotedEscapedString(output, "logger");
+	output.append(": ");
+	appendQuotedEscapedString(output, event->getLoggerName());
+	output.append(", ");
+
+	appendQuotedEscapedString(output, "message");
+	output.append(": ");
+	appendQuotedEscapedString(output, event->getMessage());
+
+	appendSerializedMDC(output, event);
+
+	appendSerializedNDC(output, event);
+
+
+	if (locationInfo)
```

Comment:
How about putting the implementation into it's custom private `formatLocationInfo` or `fmtLocInfo` or alike? Would make `format` itself a bit shorter and easier to read. Appending the first `,` could stay here to make `fmtLocInfo` fully independent.

Reply:
I tried to stay consistent with what I already have and called it appendSerializedLocationInfo.  If you'd prefer me to rename everything to shorter or otherwise different names, let me know.  Fixed.